library/qsave.pl

```diff
@@ -870,30 +882,50 @@
 save_foreign_libraries(RC, Options) :-
     option(foreign(save), Options),
     !,
-    feedback('~nFOREIGN LIBRARIES~n', []),
+    current_prolog_flag(arch, HostArch),
+    feedback('~nHOST(~w) FOREIGN LIBRARIES~n', [HostArch]),
+    save_foreign_libraries1(HostArch, RC, Options).
+
+save_foreign_libraries(RC, Options) :-
+    option(foreign(arch(Archs)), Options),
+    !,
+    forall( member(Arch, Archs),
+            ( feedback('~n~w FOREIGN LIBRARIES~n', [Arch]),
+              save_foreign_libraries1(Arch, RC, Options)
+            )).
+
+save_foreign_libraries(_, _).
+
+save_foreign_libraries1(Arch, RC, _Options) :-
     forall(current_foreign_library(FileSpec, _Predicates),
-           ( find_foreign_library(FileSpec, File, Time),
-             term_to_atom(FileSpec, Name),
+           ( find_foreign_library(Arch, FileSpec, EntryName, File, Time),
+             term_to_atom(EntryName, Name),
              zipper_append_file(RC, Name, File, [time(Time)])
            )).
-save_foreign_libraries(_, _).
 
-%!  find_foreign_library(+FileSpec, -File, -Time) is det.
+%!  find_foreign_library(+Architecture, +FileSpec, -EntryName, -File, -Time) is det.
 %
-%   Find the shared object specified by   FileSpec.  If posible, the
-%   shared object is stripped to reduce   its size. This is achieved
-%   by calling strip -o <tmp> <shared-object>. Note that the file is
-%   a Prolog tmp file and will be deleted on halt.
+%   Find the shared object specified by   FileSpec for the named
+%   `Architecture`.  `EntryName` will be the name of the file within
+%   the saved state archive. If posible, the shared object is stripped
+%   to reduce   its size. This is achieved by calling
+%   strip -o <tmp> <shared-object>.
+%   Note that (if stripped) the file is a Prolog tmp file and
+%   will be deleted on halt.
 %
 %   @bug    Should perform OS search on failure
 
-find_foreign_library(FileSpec, SharedObject, Time) :-
-    absolute_file_name(FileSpec,
-                       [ file_type(executable),
-                         access(read),
-                         file_errors(fail)
-                       ], File),
-    !,
+find_foreign_library(Arch, FileSpec, shlib(Arch,Name), SharedObject, Time) :-
+    FileSpec = foreign(Name),
+    % Call user hook to obtain shared library for architecture
+    (   ( catch( arch_find_shlib(Arch, FileSpec, File),
```

Comment:
I thought you renamed this to `arch_find_shlib`?

Reply:
Only the user hook.  I don't care too much about the internal name to change it (as it also needs to be different).

Reply:
Where do you call `arch_shlib` (the user hook)?

Reply:
Line 999

Reply:
ahh I forgot to mention, `runtime.doc` still has the old name `arch_find_shlib`